### PR TITLE
Allow custom rerouting logic.

### PR DESCRIPTION
**Pull request by wwajerowicz**
Resolves https://github.com/mapbox/mapbox-navigation-ios/issues/122

### CODE REVIEW

File: MapboxNavigation/NavigationViewController.swift [hunk 1]
```diff
@@ -280,6 +288,10 @@ public class NavigationViewController: NavigationPulleyViewController, RouteMapV
     
     func shouldReroute(notification: NSNotification) {
         let location = notification.userInfo![RouteControllerNotificationShouldRerouteKey] as! CLLocation
+
+        if navigationDelegate?.navigationViewControllerShouldReroute?(self, fromLocation: location) != nil {
```
Comment (bsudekum) [inline MapboxNavigation/NavigationViewController.swift]: Nit: want to make this a `guard` instead?

Comment (wwajerowicz) [inline MapboxNavigation/NavigationViewController.swift]: Fixed.

File: MapboxNavigation/NavigationViewController.swift [hunk 2]
```diff
@@ -21,6 +21,14 @@ public protocol NavigationViewControllerDelegate {
...
      */
     @objc optional func navigationViewController(_ navigationViewController : NavigationViewController, didArriveAt destination: MGLAnnotation)
+
+    /**
+     Called when `RouteControllerShouldReroute` is sent from `RouteController`.
+
+     If this method is unimplemented `NavigationViewController` will automatically reroute. If this method is implemented,
+     it's responsible for handling rerouting.
+    */
+    @objc optional func navigationViewControllerShouldReroute(_ : NavigationViewController, fromLocation location: CLLocation)
```
Comment (1ec5) [inline MapboxNavigation/NavigationViewController.swift]: The method should be called `navigationViewController(_:shouldRerouteFrom:)`, with an attribute of `@objc(navigationViewController:shouldRerouteFromLocation:)`.

Comment (wwajerowicz) [inline MapboxNavigation/NavigationViewController.swift]: @1ec5 yes, that method name makes much more sense.

**Pull request merged**